invoice/src/main/java/org/killbill/billing/invoice/generator/FixedAndRecurringInvoiceItemGenerator.java

```diff
@@ -176,19 +179,24 @@ void processFixedBillingEvents(final UUID invoiceId, final UUID accountId, final
         final InvoiceItemGeneratorLogger invoiceItemGeneratorLogger = new InvoiceItemGeneratorLogger(invoiceId, accountId, "fixed", log);
 
         final Iterator<BillingEvent> eventIt = events.iterator();
-        while (eventIt.hasNext()) {
-            final BillingEvent thisEvent = eventIt.next();
-            if (thisEvent.getTransitionType() == BCD_CHANGE) {
-                continue;
-            }
+        BillingEvent currentEvent = eventIt.next();
+
+        do {
+            BillingEvent nextEvent = eventIt.hasNext() ? eventIt.next() : null;
 
-            final InvoiceItem currentFixedPriceItem = generateFixedPriceItem(invoiceId, accountId, thisEvent, targetDate, currency, invoiceItemGeneratorLogger, internalCallContext);
-            if (!isSameDayAndSameSubscription(prevItem, thisEvent, internalCallContext) && prevItem != null) {
-                proposedItems.add(prevItem);
+            if (currentEvent.getTransitionType() != BCD_CHANGE) {
+                final BillingEvent adjustedNextEvent = (currentEvent.getSubscriptionId() == (nextEvent != null ? nextEvent.getSubscriptionId() : null)) ? nextEvent : null;
```

Comment:
The double ternary statement is hard to follow but I agree it should work... except that checking for equality of UUID should use `equals()` and not `==` but it seems like this bad code is already [there for recurring items](https://github.com/killbill/killbill/blob/master/invoice/src/main/java/org/killbill/billing/invoice/generator/FixedAndRecurringInvoiceItemGenerator.java#L154). So I suggest:

1. Fix those in this branch 
2. Independently, let's open a small PR to fix the recurring case in `master` - which mean we may have a conflict ton resolve when me merge.